### ISSUE TITLE
Retrieving User's Identity fails with ObjectDisposedException

### ISSUE THREAD

**Issue by vcaraulean** [aspnet/Logging#543]
Looking for guidance here, this problem will most likely affect any custom logging implementations.

Consider next implementation of `ILogger.Log` method:

    public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
    {
        var message = formatter(state, exception);
        var user = _httpContextAccessor.HttpContext?.User?.Identity.Name;
        Debug.WriteLine($"================= {user ?? "<n/a>"}: {message}");
    }

After few successful calls, retrieval of user's Identity.Name fails. 

Log message at the moment of failure:

    "Connection id \"0HL1QFAEBG2MK\" completed keep alive response."

Exception details:

    System.ObjectDisposedException was unhandled by user code
    HResult=-2146232798
    Message=Safe handle has been closed
    ObjectName=""
    Source=mscorlib
    StackTrace:
        at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
        at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
        at Microsoft.Win32.Win32Native.GetTokenInformation(SafeAccessTokenHandle TokenHandle, UInt32 TokenInformationClass, SafeLocalAllocHandle TokenInformation, UInt32 TokenInformationLength, UInt32& ReturnLength)
        at System.Security.Principal.WindowsIdentity.GetTokenInformation(SafeAccessTokenHandle tokenHandle, TokenInformationClass tokenInformationClass)
        at System.Security.Principal.WindowsIdentity.get_User()
        at System.Security.Principal.WindowsIdentity.GetName()
        at System.Security.Principal.WindowsIdentity.get_Name()
        at WebApplication2.CustomLogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter) in C:\Dev\Temp\NLog.Web.AspNetCore.Issue83\src\WebApplication2\CustomLogger.cs:line 21
        at Microsoft.Extensions.Logging.Logger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)

SDK:

    "sdk": {
      "version": "1.0.0-preview2-1-003177"
    }

Is there a safe way to retrieve `User.Identity.Name` or to detect if this information not available? Which will be the best way for logging frameworks to safely extract this info?

Catching the exception would be one way to do it. But it's ugly and smells...

**Comment by vcaraulean** [comment]
Any recommendations for dealing with this issue?

**Comment by glennc** [comment]
@Tratcher Can you help with this?

**Comment by Tratcher** [comment]
You're using Windows auth forwarded from IIS, correct? The IISIntegration middleware disposes of the identity at the end of the request. Maybe it should also clear HttpContext.User at this point.

That said, I'm surprised the HttpContextAccessor has not been cleared at this point. We should check when this log statement occures in kestrel.

**Comment by Tratcher** [comment]
@cesarbs can you check when Kestrel logs `"Connection id \"0HL1QFAEBG2MK\" completed keep alive response."`. Is it ever before the request pipeline exits? Maybe for a Connection-Length response?

**Comment by cesarbs** [comment]
@Tratcher It's always logged before the `HttpContext` is disposed.

**Comment by 304NotModified** [comment]
Isn't this a bug? I would expect we never should get a `System.ObjectDisposedException` and also throwing an exception from a property is not advised?

**Comment by 304NotModified** [comment]
polite bump

**Comment by muratg** [comment]
Removing from Discussions milestone for a re-triage.

**Comment by jchannon** [comment]
I am also seeing this with the below code and stack trace:

```
                    var principalUser = context.User as WindowsPrincipal;

                    if (principalUser != null)
                    {
                        var loginId = principalUser.Identity.Name;
                    }
```

```
2017-06-30 22:07:47.442073  RequestId : 800142e7-000e-d500-b63f-84710c7967bb Component : VQPortal Method : OWIN Pipeline Message : System.ObjectDisposedException: Safe handle has been closed
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Interop.mincore.LsaLookupSids(SafeLsaPolicyHandle handle, Int32 count, IntPtr[] sids, SafeLsaMemoryHandle& referencedDomains, SafeLsaMemoryHandle& names)
   at System.Security.Principal.SecurityIdentifier.TranslateToNTAccounts(IdentityReferenceCollection sourceSids, Boolean& someFailed)
   at System.Security.Principal.SecurityIdentifier.Translate(IdentityReferenceCollection sourceSids, Type targetType, Boolean forceSuccess)
   at System.Security.Principal.SecurityIdentifier.Translate(Type targetType)
   at System.Security.Principal.WindowsIdentity.<GetName>b__42_0()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Security.Principal.WindowsIdentity.RunImpersonatedInternal(SafeAccessTokenHandle token, Action action)
   at System.Security.Principal.WindowsIdentity.GetName()
```

This is on `"Microsoft.AspNetCore.Server.WebListener": "1.1.0"` with
```
new WebHostBuilder()
                .UseWebListener(options =>
                {
                        options.ListenerSettings.Authentication.Schemes = AuthenticationSchemes.Negotiate;
                })
```

**Comment by Tratcher** [comment]
Debug notes...
Full stack for the IIS Express Windows Auth scenario:
```
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown	Non-user code. Skipped loading symbols.
 	System.Security.Principal.Windows.dll!System.Security.Principal.WindowsIdentity.RunImpersonatedInternal(Microsoft.Win32.SafeHandles.SafeAccessTokenHandle token, System.Action action)	Unknown	Non-user code. Skipped loading symbols.
 	System.Security.Principal.Windows.dll!System.Security.Principal.WindowsIdentity.RunImpersonated(Microsoft.Win32.SafeHandles.SafeAccessTokenHandle safeAccessTokenHandle, System.Action action)	Unknown	Non-user code. Skipped loading symbols.
 	System.Security.Principal.Windows.dll!System.Security.Principal.WindowsIdentity.GetName()	Unknown	Non-user code. Skipped loading symbols.
 	System.Security.Principal.Windows.dll!System.Security.Principal.WindowsIdentity.Name.get()	Unknown	Non-user code. Skipped loading symbols.
 	IISSample.dll!IISSample.CustomConsoleLogger.Log<object>(Microsoft.Extensions.Logging.LogLevel logLevel, Microsoft.Extensions.Logging.EventId eventId, object state, System.Exception exception, System.Func<object, System.Exception, string> formatter) Line 52	C#	Symbols loaded.
 	Microsoft.Extensions.Logging.dll!Microsoft.Extensions.Logging.Logger.Log<object>(Microsoft.Extensions.Logging.LogLevel logLevel, Microsoft.Extensions.Logging.EventId eventId, object state, System.Exception exception, System.Func<object, System.Exception, string> formatter)	Unknown	Non-user code. Skipped loading symbols.
 	Microsoft.Extensions.Logging.Abstractions.dll!Microsoft.Extensions.Logging.Logger<Microsoft.AspNetCore.Hosting.Internal.WebHost>.Log<object>(Microsoft.Extensions.Logging.LogLevel logLevel, Microsoft.Extensions.Logging.EventId eventId, object state, System.Exception exception, System.Func<object, System.Exception, string> formatter)	Unknown	Non-user code. Skipped loading symbols.
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.HostingApplicationDiagnostics.LogRequestFinished(Microsoft.AspNetCore.Http.HttpContext httpContext, long startTimestamp, long currentTimestamp)	Unknown	Non-user code. Skipped loading symbols.
>	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.HostingApplication.DisposeContext(Microsoft.AspNetCore.Hosting.Internal.HostingApplication.Context context, System.Exception exception)	Unknown	Non-user code. Skipped loading symbols.
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Frame<Microsoft.AspNetCore.Hosting.Internal.HostingApplication.Context>.ProcessRequestsAsync()	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.OutputAsyncCausalityEvents.AnonymousMethod__0()	Unknown	Non-user code. Skipped loading symbols.
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.LoggingThreadPool.RunAction.AnonymousMethod__3_0(object o)	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()	Unknown	Non-user code. Skipped loading symbols.
```
The log statement is coming from hosting:
https://github.com/aspnet/Hosting/blob/456957bf57901da81d943166fb7f0b4725e7b45e/src/Microsoft.AspNetCore.Hosting/Internal/HostingApplication.cs#L53
Which is just before HttpContextAccessor gets cleared:
https://github.com/aspnet/HttpAbstractions/blob/594f55947f4c1d0a9d3122e3f39bcfa81199b12a/src/Microsoft.AspNetCore.Http/HttpContextFactory.cs#L54

Kestrel calls DisposeContext after it has called FireOnCompleted.
https://github.com/aspnet/KestrelHttpServer/blob/6584a8b5fdaba0a79c52b981cac88472bcc92d1a/src/Kestrel.Core/Internal/Http/FrameOfT.cs#L187
https://github.com/aspnet/KestrelHttpServer/blob/6584a8b5fdaba0a79c52b981cac88472bcc92d1a/src/Kestrel.Core/Internal/Http/FrameOfT.cs#L136
Which is what IISIntegration uses to dispose the user.
https://github.com/aspnet/IISIntegration/blob/f6329c4dc7e1fc2e0000d646ded8f81b85ad6296/src/Microsoft.AspNetCore.Server.IISIntegration/AuthenticationHandler.cs#L53

HttpSys:
HttpSys also disposes the user before calling DisposeContext:
https://github.com/aspnet/HttpSysServer/blob/17852e521768c2a1e2626d6d67e7300fe2e5c201/src/Microsoft.AspNetCore.Server.HttpSys/RequestProcessing/Request.cs#L320
https://github.com/aspnet/HttpSysServer/blob/17852e521768c2a1e2626d6d67e7300fe2e5c201/src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs#L205-L206

So everything is behaving 'correctly', but there's a window where the user has been disposed but IHttpContextAccessor has not yet been cleared, and Hosting always logs within this window.

Not causal, but relevant to any changes we may make here: There are also some subtle ordering differences between Kestrel and HttpSys. E.g. Kestrel fires OnCompleted before it Calls DisposeContext, where HttpSys calls it afterwards. Calling OnCompleted first is perhaps more correct as the user may have closed over the HttpContext.

Possible fixes (most of these will differ between the HttpSys and IIS scenarios due to which component creates the user):
IISIntegration: Close over the HttpContext and clear the user on Dispose.
https://github.com/aspnet/IISIntegration/blob/f6329c4dc7e1fc2e0000d646ded8f81b85ad6296/src/Microsoft.AspNetCore.Server.IISIntegration/AuthenticationHandler.cs#L53
HttpSys: Reorder the Dispose and DisposeContext calls so the low level resources are disposed after the higher level HttpContext.  We should also move the OnCompleted call to run first.
https://github.com/aspnet/HttpSysServer/blob/17852e521768c2a1e2626d6d67e7300fe2e5c201/src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs#L205-L210

**Comment by jchannon** [comment]
Fixed?

**Comment by Tratcher** [comment]
Yes, see the linked PRs.

**Comment by jchannon** [comment]
Thanks

**Comment by jchannon** [comment]
Will this be out in 2.1 or as part of an update to current 2.0 SDK?

**Comment by Tratcher** [comment]
This was fixed for 2.1
